enhancements/upgrade-readiness.md

```diff
@@ -0,0 +1,426 @@
+status: implementable
+see-also:
+  - "N/A"  
+replaces:
+  - "N/A"
+superseded-by:
+  - "N/A"
+---
+
+# operator-status
+
+## Release Signoff Checklist
+
+- [x] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [x] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+
+## Summary
+
+The goal of this enhancement is to allow operators managed by the [Operator Lifecycle Manager (OLM)](https://github.com/operator-framework/operator-lifecycle-manager) to communicate complex and potentially dangerous conditions and states to OLM. OLM will modify its behavior based on this information to protect the cluster.
+
+## Motivation
+
+OLM currently infers the state of an operator by inspecting the Kubernetes resources that define the operator. OLM does not provide operators with the means to communicate complex states that cannot be inferred without operator specific knowledge.
+
+Operators managed by OLM would benefit from the ability to communicate their state directly to OLM through a dedicated channel. Operators could communicate complex conditions and influence how OLM manages the operator, dramatically improving user experience. For example, an operator could communicate to OLM that the operator should not be upgraded or uninstalled while a critical process is being executed. Consider the following scenario based on OLM today:
+
+>An operator installed with automatic updates is executing a critical process that could place the cluster in a bad state if interrupted. While this critical process is running, a new version of the operator becomes available. Since the operator was installed with automatic updates, OLM upgrades the operator without allowing the critical process to finish and places the cluster in an unrecoverable state.
+
+OLM can provide both operator authors and cluster admins a significantly better user experience by creating a channel for the operator to communicate state that OLM cannot infer otherwise.
+
+### Goals
+
+- Provide operators with the means to communicate complex states that OLM cannot infer.
+- Provide operators with the means to inform OLM that the operator should not be upgraded or uninstalled.
+- Update OLM not to upgrade operators if the operator communicates that it should not be upgraded.
+- Preserve existing behavior should the operator not communicate its upgradeable status.
+
+### Non-Goals
+
+- Support operator version rollback.
+- Provide operators with the means to cancel an upgrade that is in process.
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+As an operator author, I want my operator to be able to communicate complex states to OLM that it cannot infer otherwise.
+
+#### Story 2 (core requirement)
+
+As an operator author, I want my operator to be able to communicate that OLM should not attempt to upgrade or uninstall my operator until I tell it otherwise.
+
+## Design Details
+
+### The OperatorStatus CRD
```

Comment:
One question I have is whether this CRD is namespace scoped or cluster scoped? How do we handle cases where we have multiple versions of the same operator installed in the same cluster, like the AWS connector operator? 

If there is a 1-1 mapping of the CR to each operator, what happens when some operators are ready to update and others are not. It's an edge case but worth thinking about

Reply:
This question was directed at the previous design, but I want to address it because it is still relevant to the existing approach.

The plan was for the CRD introduced by OLM to be namespaced. If an operator is installed in multiple namespaces on a cluster and one of the operators are reporting that the operator cannot be resolved, the operator will not upgrade without cluster admin intervention. This is by design, OLM cannot ignore an operator that reports that it cannot be upgraded.

Reply:
Although this question was brought up against the previous design, it is still a great question!

> One question I have is whether this CRD is namespace scoped or cluster scoped

The proposed CRD would have been namespaced.

> If there is a 1-1 mapping of the CR to each operator, what happens when some operators are ready to update and others are not. It's an edge case but worth thinking about

Given that the CRD is namespaced and each installation of the operator is allowed to communicate their upgrade readiness I do not believe this would be an issue.